add IMEX-BDF2-type time stepping

boyceg opened the pull request:
Adding support for BDF2-like time stepping. The goal is to enable "consistent" time stepping for IIM.

<!--
This template should be included in all pull requests. Items in the list should
either be completed by the original author or explicitly dismissed by one of the
IBAMR principal developers.

IBAMR is a community effort and it wouldn't exist without people contributing
code. Thanks in advance for helping to make IBAMR better!
-->

### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [ ] Was clang-format (i.e., `make indent`) run? For more information see
      `scripts/formatting/README.md`.
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

boyceg commented:
The implementation is intended to correspond to VSSBDF2 from [this paper](https://steveruuth.org/wp-content/uploads/2020/10/vsimex_Wang_Ruuth_jcm.pdf).

boyceg commented:
One thing that I know that I need to fix is regridding the old velocity field.

boyceg commented:
[Here is a paper](https://www.sciencedirect.com/science/article/pii/S0021999119303237) looking at errors in the initial time step. The recommendation is to use something like TR-BDF2 in the first time step.

boyceg commented:
@abarret I am not getting perfect second-order convergence rates; I wonder if it is start-up error?

abarret commented on src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp:
```diff
@@ -2287,27 +2384,43 @@ INSStaggeredHierarchyIntegrator::reinitializeOperatorsAndSolvers(const double cu
     PoissonSpecifications P_problem_coefs(d_object_name + "::P_problem_coefs");
     P_problem_coefs.setCZero();
-    P_problem_coefs.setDConstant(rho == 0.0 ? -1.0 : -1.0 / rho);
+    P_problem_coefs.setDConstant(rho == 0.0 ? -1.0 : -1.0 / (K1 * rho));
```
P is not timestepped.
```suggestion
    P_problem_coefs.setDConstant(rho == 0.0 ? -1.0 : -1.0 / rho);
```

boyceg replied:
Here I have to admit that I have forgotten why we need to include `rho` in the pressure setup.

abarret replied:
Oh hmmm... I think I misunderstood what this was being used for.

boyceg replied:
Specifically, at least for the test I am running, this seems to make no difference for `TRAPEZOIDAL_RULE` but to shave one iteration off for `BDF2`. Let me see if I can figure out whether we care about this or not in practice.

boyceg replied:
This is a little obscure, but the projection preconditioner is derived from
```
    //    rho (U - U^*) + G Phi = 0
    //    -D U = F_P
```
and then figuring out how `Phi` relates to `P`. I think that for a BDF scheme, we need to do instead
```
    //    K1 rho (U - U^*) + G Phi = 0
    //    -D U = F_P
```
where `K1` is this factor that is hitting `U^{n+1}` in the BDF time stepping scheme.

boyceg commented:
I seem to be getting essentially perfect 2nd-order accuracy with the uniform grid version of `examples/navier_stokes/ex0` with periodic boundary conditions.

abarret commented on src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp:
```diff
@@ -1440,15 +1490,28 @@ INSStaggeredHierarchyIntegrator::setupSolverVectors(const Pointer<SAMRAIVectorRe
-            d_hier_sc_data_ops->linearSum(N_idx, 1.0 + 0.5 * omega, N_idx, -0.5 * omega, d_N_old_current_idx);
+            const double omega = getTimeStepSizeRatio();
+            double beta1, beta2;
+            if (is_bdf_time_stepping_type(d_viscous_time_stepping_type))
+            {
+                beta1 = 1.0 + omega;
+                beta2 = -omega;
+            }
```
I don't think this is true for `BACKWARD_EULER`. The coefficients depend on the particular BDF scheme.

boyceg replied:
Don't you want the RHS to be at the end of the time step at least for any BDF with order >= 2?

abarret replied:
Yes, but `BACKWARD_EULER` is not order >= 2. I guess if you are using first order viscous time stepping, using first order convective time stepping is fine too?

boyceg replied:
Well, what is done now is to use time-step centered convective time stepping with `BACKWARD_EULER`.

I can drop `BACKWARD_EULER` from the list of BDF time stepping schemes if you would prefer that. 😁

boyceg replied:
I have removed `BACKWARD_EULER` from the list of BDF time stepping schemes for now. Maybe at some point we can rationalize this stuff.

abarret commented on src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp:
```diff
@@ -1464,23 +1527,57 @@ INSStaggeredHierarchyIntegrator::setupSolverVectors(const Pointer<SAMRAIVectorRe
     if (d_F_fcn)
     {
-        d_F_fcn->setDataOnPatchHierarchy(d_F_scratch_idx, d_F_var, d_hierarchy, half_time);
+        double data_time;
+        if (is_bdf_time_stepping_type(d_viscous_time_stepping_type))
+        {
+            data_time = new_time;
+        }
+        else
+        {
+            data_time = half_time;
+        }
+        d_F_fcn->setDataOnPatchHierarchy(d_F_scratch_idx, d_F_var, d_hierarchy, data_time);
```
This might be a problem for body forcing terms that depend on advected quantities, for example I don't know if the complex fluid solver is set up to correctly handle evaluations at `new_time`.

I have switched over to using AB2 for body forces.

boyceg replied:
This change is actually the primary reason that I am doing this --- we want to be able to use the end-time step forcing for IIM.

abarret replied:
Yes, but we need a better way to inform force functions that the approximation should be using data at `t^{n+1}` instead of `t^{n+1/2}`. For example, this always uses `t^{n+1/2}` regardless of time stepping scheme, which for BDF2 will be first order.
https://github.com/IBAMR/IBAMR/blob/2b40ae33badfd03f2759c8999705cf91060da548/examples/navier_stokes/ex5/BoussinesqForcing.cpp#L76-L86

boyceg replied:
Well, here the implementation (probably by me) is ignoring `data_time`....!

boyceg replied:
There is a hook in the API that callers can use to inform when things should be evaluated. If the called function cannot evaluate at that time, it should just emit an error message and abort.

abarret replied:
I agree, but there is not a great API to figure out which time point `data_time` corresponds to. But this is an issue with the use of `CartGridFunction`, not the time stepping scheme.

boyceg replied:
I've tried to capture this issue and a possible approach in #1709.

boyceg replied:
Also #1710. It would be nice to be able to handle time extrapolation in a more uniform manner, so that (e.g.) we could easily allow for either evaluation at the end of the time step or extrapolation to the end of the time step in places like this.

abarret reviewed (changes requested):
Some code clarifications. We need to add tests. I think I will have time tomorrow or early next week to do that if you want me to add them.

boyceg commented:
> Some code clarifications. We need to add tests. I think I will have time tomorrow or early next week to do that if you want me to add them.

If you have the time and interest to take it over, that would be great. I am now working on a couple of grant applications.

boyceg commented on include/ibamr/IBFEMethod.h:
```diff
@@ -625,6 +626,11 @@ class IBFEMethod : public FEMechanicsBase, public IBStrategy
+    /*!
+     * Indicate that multistep time stepping will be used.
+     */
+    void setUseMultistepTimeStepping(int n_steps = 1) override;
```
Added some documentation and changed the variable name to `n_previous_steps`.

boyceg commented:
@abarret can you take a look and see if this has addressed your concerns?

abarret commented:
> @abarret can you take a look and see if this has addressed your concerns?

This looks good to me. I have some tests, may I push them to this branch?

boyceg commented:
> > @abarret can you take a look and see if this has addressed your concerns?
> 
> This looks good to me. I have some tests, may I push them to this branch?

That would be wonderful.

abarret reviewed (approved):
I've added two tests for BDF2.

boyceg commented:
@drwells or @amneetb, does one of you want to approve before we merge since @abarret and I are now effectively collaborating on this PR? 😁

boyceg commented:
@abarret do you feel like squashing?

amneetb commented:
I can approve but I am not listed as a reviewer on this PR.

abarret commented:
Squashed
